modules/3d/src/octree.cpp

```diff
@@ -0,0 +1,507 @@
+    Ptr <OctreeNode> rootNode = nullptr;
+    //! Max depth of the Octree. And depth must be greater than zero
+    int maxDepth;
+    //! The size of the cube of the .
+    double size;
+    //! The origin coordinate of root node.
+    Point3f origin;
+};
+
+Octree::Octree() : p(new Impl)
+{
+    p->maxDepth = -1;
+    p->size = 0;
+    p->origin = Point3f(0,0,0);
+}
+
+Octree::Octree(int _maxDepth, double _size, const Point3f& _origin ) : p(new Impl)
+{
+    p->maxDepth = _maxDepth;
+    p->size = _size;
+    p->origin = _origin;
+}
+
+Octree::Octree(int _maxDepth, const std::vector<Point3f>& _pointCloud) : p(new Impl)
+{
+    CV_Assert( _maxDepth > -1 );
+    p->maxDepth = _maxDepth;
+    convertFromPointCloud(_pointCloud);
+}
+
+Octree::Octree(int _maxDepth) : p(new Impl)
+{
+    p->maxDepth = _maxDepth;
+    p->size = 0;
+    p->origin = Point3f(0,0,0);
+}
+
+Octree::~Octree(){}
+
+void Octree::insertPoint(const Point3f& point)
+{
+    if(p->rootNode == nullptr)
+    {
+        p->rootNode = new OctreeNode( 0, p->size, p->origin, -1);
+    }
+
+    insertPointRecurse(p->rootNode, point, p->maxDepth);
+}
+
+bool Octree::convertFromPointCloud(const std::vector<Point3f> &pointCloud)
+{
+    CV_Assert( p->maxDepth > -1 && !pointCloud.empty());
+
+    if(pointCloud.empty())
+        return false;
+
+    Point3f maxBound(pointCloud[0]);
+    Point3f minBound(pointCloud[0]);
+    Point3f center, temp;
+
+    // Find center coordinate of PointCloud data.
+    for(size_t idx = 0; idx <pointCloud.size(); idx++)
+    {
+        maxBound.x = max(pointCloud[idx].x, maxBound.x);
+        maxBound.y = max(pointCloud[idx].y, maxBound.y);
+        maxBound.z = max(pointCloud[idx].z, maxBound.z);
+
+        minBound.x = min(pointCloud[idx].x, minBound.x);
+        minBound.y = min(pointCloud[idx].y, minBound.y);
+        minBound.z = min(pointCloud[idx].z, minBound.z);
+    }
+
+    center = (maxBound + minBound) / 2.0f;
+
+    temp = center - minBound;
+    float halfSize = std::max(temp.x, std::max(temp.y, temp.z));
+    p->origin = center - Point3f(halfSize, halfSize, halfSize);
+    p->size = 2 * halfSize;
+
+    // Insert every point in PointCloud data.
+    for(size_t idx = 0; idx < pointCloud.size(); idx++ )
+    {
+        insertPoint(pointCloud[idx]);
+    }
+
+    return true;
+}
+
+void Octree::setMaxDepth(int _maxDepth)
+{
+    if(_maxDepth )
+    p->maxDepth = _maxDepth;
+}
+
+void Octree::setSize(double _size)
+{
+    p->size = _size;
+};
+
+void Octree::setOrigin(const Point3f& _origin)
+{
+    p->origin = _origin;
+}
+
+void Octree::clear()
+{
+    if(p->rootNode != nullptr)
+    {
+        p->rootNode.release();
+    }
+
+    p->size = 0;
+    p->maxDepth = -1;
+    p->origin = Point3f (0,0,0); // origin coordinate
+}
+
+bool Octree::empty()
+{
+    return p->rootNode == nullptr;
+}
+
+Ptr<OctreeNode> index(const Point3f& point, Ptr<OctreeNode>& node)
+{
+    if(node == nullptr)
+    {
+        return nullptr;
+    }
```

Comment:
1. `node == nullptr` comparison is not efficient. Use `operator bool` instead.
2. `return nullptr;` is not efficient.
3. reduce amount of `node->` dereferences.

```
Ptr<OctreeNode> index(const Point3f& point, Ptr<OctreeNode>& node_)
{
    if (!node_)
    {
        return Ptr<OctreeNode>();
    }

    OctreeNode& node = *node_;

    ...
}
```

---

Probably you don't need `Ptr<>` in this implementation.

```
Ptr<OctreeNode> index(const Point3f& point, const OctreeNode& node)
```